cmd/buildah/images.go

```diff
@@ -299,46 +304,95 @@ func outputImages(ctx context.Context, systemContext *types.SystemContext, store
-							CreatedAt:    units.HumanDuration(time.Since((createdTime))) + " ago",
-							Size:         formattedSize(size),
-							ReadOnly:     image.ReadOnly,
-						})
-					// We only want to print each id once
-					break outer
-				}
-				params := imageOutputParams{
-					Tag:          tag,
-					ID:           imageID,
-					Name:         name,
-					Digest:       digest,
-					CreatedAtRaw: createdTime,
-					CreatedAt:    units.HumanDuration(time.Since((createdTime))) + " ago",
-					Size:         formattedSize(size),
-					ReadOnly:     image.ReadOnly,
+		var imageReposAndTags [][2]string
+		var imageDigests []string
+		for _, imageDigest := range image.Digests {
+			imageDigests = append(imageDigests, imageDigest.String())
+		}
+
+		for _, name := range image.Names {
+			if name == "" {
+				logrus.Warnf("Found image with empty name")
+				continue
+			}
+			named, err := reference.ParseNormalizedNamed(name)
+			if err != nil {
+				logrus.Warnf("Error parsing name %q: %v", name, err)
+				continue
+			}
+			if name != named.String() {
+				logrus.Debugf("Image name %q wasn't already in its normalized form (%q).", name, named.String())
+			}
+
+			if !matchesReference(name, argName) {
+				continue
+			}
+			found = true
+
+			if digested, ok := named.(reference.Digested); ok {
+				digest := digested.Digest()
+				digestPresent := false
+				for _, imageDigest := range imageDigests {
+					if imageDigest == digest.String() {
+						digestPresent = true
+					}
 				}
-				imagesParams = append(imagesParams, params)
-				if opts.quiet {
-					// We only want to print each id once
-					break outer
+				if !digestPresent {
+					imageDigests = append(imageDigests)
 				}
 			}
+
+			if !matchesFilter(ctx, store, image, name, filters) {
+				continue
+			}
+			filterMatched = true
+
+			if tagged, ok := named.(reference.Tagged); ok {
+				imageReposAndTags = append(imageReposAndTags, [2]string{named.Name(), tagged.Tag()})
+			} else {
+				imageReposAndTags = append(imageReposAndTags, [2]string{named.Name(), ""})
+			}
+		}
+		if len(image.Names) == 0 && matchesFilter(ctx, store, image, "", filters) {
+			filterMatched = true
+		}
+		if !filterMatched {
+			continue
+		}
+
+		if opts.json {
+			jsonImages = append(jsonImages, jsonImage{
+				ID:           image.ID,
+				Names:        image.Names,
+				Digest:       string(image.Digest),
+				Digests:      imageDigests,
+				CreatedAtRaw: createdTime,
+				CreatedAt:    units.HumanDuration(time.Since((createdTime))) + " ago",
+				Size:         formattedSize(size),
+				ReadOnly:     image.ReadOnly,
+			})
+			continue
+		}
+		if len(imageReposAndTags) == 0 {
+			imageReposAndTags = [][2]string{{"", ""}}
+		}
+		for _, imageRepoAndTag := range imageReposAndTags {
+			imagesParams = append(imagesParams, imageOutputParams{
+				ID:           imageID,
+				Name:         imageRepoAndTag[0],
+				Tag:          imageRepoAndTag[1],
+				Digest:       string(image.Digest),
```

Comment:
This will output the primary `image.Digest` even if `imageRepoAndTag` came from a digested reference with a different digest — even if there are multiple digested references with different digests, they will all report the same `Digest:` value. Is that acceptable?